std/bitmanip.d

```diff
@@ -799,18 +801,157 @@ private:
+
+    Params:
+        v = Source array. `v.length` must be a multple of `size_t.sizeof`.
+        numbits = Number of bits to be mapped from the source array, i.e.
+                  length of the created `BitArray`.
+    */
+    this(void[] v, size_t numbits) @nogc nothrow pure
+    in
+    {
+        assert(numbits <= v.length * 8);
+        assert(v.length % size_t.sizeof == 0);
+    }
+    do
+    {
+        _ptr = cast(size_t*) v.ptr;
+        _len = numbits;
+        if (endBits)
+        {
+            // Need to mask away extraneous bits from v.
+            _ptr[dim - 1] &= endMask;
+        }
+    }
+
+    ///
+    @system unittest
+    {
+        import std.algorithm.comparison : equal;
+
+        auto a = BitArray([1, 0, 0, 1, 1]);
+
+        // Inverse of the cast.
+        auto v = cast(void[]) a;
+        auto b = BitArray(v, a.length);
+
+        assert(b.length == 5);
+        assert(b.bitsSet.equal([0, 3, 4]));
+
+        // a and b share the underlying data.
+        a[0] = 0;
+        assert(b[0] == 0);
+        assert(a == b);
+    }
+
+    ///
+    @system unittest
+    {
+        import std.algorithm.comparison : equal;
+
+        size_t[] source = [0b1100, 0b0011];
+        auto sbits = size_t.sizeof * 8;
+        auto ba = BitArray(source, source.length * sbits);
+        // The least significant bit in each unit is this unit's starting bit.
+        assert(ba.bitsSet.equal([2, 3, sbits, sbits + 1]));
+    }
+
+    ///
+    @system unittest
+    {
+        size_t[] source = [1, 0b101, 3, 3424234, 724398, 230947, 389492];
+        auto sbits = size_t.sizeof * 8;
+        auto ba = BitArray(source, source.length * sbits);
+        foreach (n; 0 .. source.length * sbits)
+        {
+            auto nth_bit = cast(bool) (source[n / sbits] & (1L << (n % sbits)));
+            assert(ba[n] == nth_bit);
+        }
+        // Mapping only part of the array.
+        import std.algorithm.comparison : equal;
+
+        auto bc = BitArray(source, sbits + 1);
+        assert(bc.bitsSet.equal([0, sbits]));
+        // The unmapped bits from the final word have been cleared.
+        assert(source[1] == 1);
+    }
+
+    // Deliberately undocumented: raw initialization of bit array.
+    this(size_t len, size_t* ptr) @nogc nothrow pure
+    {
+        _len = len;
+        _ptr = ptr;
+    }
+
+    /**
+    Returns the dimension i.e. a number of native words backing this `BitArray`.
+
+    Technically, this is the length of the underlying array storing bits, which
+    is equal to `dim == 8 * ceil(length / size_t.sizeof)`, as bits are
+    packed into `size_t` units.
+    */
+    @property size_t dim() const @nogc nothrow pure @safe
     {
         return lenToDim(_len);
     }
 
-    /**********************************************
-     * Gets the amount of bits in the $(D BitArray).
-     */
-    @property size_t length() const @nogc pure nothrow @safe
+    /**
+    Returns the number of bits in the `BitArray`.
```

Comment:
"Returns:"

Reply:
I have one issue with that, because when I specify just `Returns:` without any other description, ddox is going to generate an empty description:
![ddox_missing_desc](https://user-images.githubusercontent.com/1181809/35772477-b5b123de-093f-11e8-8c08-ab85af6ad0e0.png)
Is this something I should care about? If yes, what is the best way to avoid writing the same thing in both the description and the `Returns` section?

Reply:
Opening an issue at Ddox: https://github.com/rejectedsoftware/ddox/issues
Our main documentation engine is still Ddoc though, there are still many visual glitches like this one with Ddox.

Reply:
Seems good, I'll change it. I found out this issue is already reported: https://github.com/rejectedsoftware/ddox/issues/172.